storm-core/src/jvm/backtype/storm/spout/CheckpointSpout.java

```diff
@@ -0,0 +1,280 @@
+import static backtype.storm.spout.CheckPointState.State.COMMITTING;
+import static backtype.storm.spout.CheckPointState.State.PREPARING;
+
+/**
+ * Emits checkpoint tuples which is used to save the state of the {@link backtype.storm.topology.IStatefulComponent}
+ * across the topology. If a topology contains Stateful bolts, Checkpoint spouts are automatically added
+ * to the topology. There is only one Checkpoint task per topology.
+ * <p/>
+ * Checkpoint spout stores its internal state in a {@link KeyValueState}. The state transitions are as follows.
+ * <p/>
+ * <pre>
+ *                  ROLLBACK(tx2)
+ *               <-------------                  PREPARE(tx2)                     COMMIT(tx2)
+ * COMMITTED(tx1)-------------> PREPARING(tx2) --------------> COMMITTING(tx2) -----------------> COMMITTED (tx2)
+ *
+ *
+ * </pre>
+ */
+public class CheckpointSpout extends BaseRichSpout {
+    private static final Logger LOG = LoggerFactory.getLogger(CheckpointSpout.class);
+
+    public static final String CHECKPOINT_STREAM_ID = "$checkpoint";
+    public static final String CHECKPOINT_COMPONENT_ID = "$checkpointspout";
+    public static final String CHECKPOINT_FIELD_TXID = "txid";
+    public static final String CHECKPOINT_FIELD_ACTION = "action";
+    public static final String CHECKPOINT_ACTION_PREPARE = "prepare";
+    public static final String CHECKPOINT_ACTION_COMMIT = "commit";
+    public static final String CHECKPOINT_ACTION_ROLLBACK = "rollback";
+    public static final String CHECKPOINT_ACTION_INITSTATE = "initstate";
+
+    private static final String TX_STATE_KEY = "__state";
+    private static final int DEFAULT_CHECKPOINT_INTERVAL = 1000; // every sec
+
+    private TopologyContext context;
+    private SpoutOutputCollector collector;
+    private long lastCheckpointTs;
+    private int checkpointInterval;
+    private boolean recoveryStepInProgress;
+    private boolean checkpointStepInProgress;
+    private boolean recovering;
+    private KeyValueState<String, CheckPointState> checkpointState;
+
+    @Override
+    public void open(Map conf, TopologyContext context, SpoutOutputCollector collector) {
+        open(context, collector, loadCheckpointInterval(conf), loadCheckpointState(conf, context));
+    }
+
+    // package access for unit test
+    void open(TopologyContext context, SpoutOutputCollector collector,
+              int checkpointInterval, KeyValueState<String, CheckPointState> checkpointState) {
+        this.context = context;
+        this.collector = collector;
+        this.checkpointInterval = checkpointInterval;
+        this.checkpointState = checkpointState;
+        lastCheckpointTs = 0;
+        recoveryStepInProgress = false;
+        checkpointStepInProgress = false;
+        recovering = true;
+    }
+
+    @Override
+    public void nextTuple() {
+        if (shouldRecover()) {
+            LOG.debug("In recovery");
+            handleRecovery();
+            startProgress();
```

Comment:
Actually it can't be moved out since its in the `elseif`. If nextTuple is invoked while its in the middle of recovering/checkpointing, it does not do anything. I will add a final else and put logs/comment.